processor/lsmintervalprocessor/processor.go

```diff
@@ -0,0 +1,388 @@
+	// processor context is cancelled ensuring export goroutine will be noop.
+	if p.exportStopped != nil {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("failed to shutdown due to context timeout while waiting for export to stop: %w", ctx.Err())
+		case <-p.exportStopped:
+		}
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	// Ensure all data in the database is exported
+	if p.db != nil {
+		p.logger.Info("exporting all data before shutting down")
+		if p.batch != nil {
+			if err := p.batch.Commit(p.wOpts); err != nil {
+				return fmt.Errorf("failed to commit batch: %w", err)
+			}
+			if err := p.batch.Close(); err != nil {
+				return fmt.Errorf("failed to close batch: %w", err)
+			}
+			p.batch = nil
+		}
+
+		var errs []error
+		for _, ivl := range p.intervals {
+			// At any particular time there will be 1 export candidate for
+			// each aggregation interval. We will align the end time and
+			// process each of these.
+			to := p.processingTime.Truncate(ivl).Add(ivl)
+			if err := p.export(ctx, to); err != nil {
+				errs = append(errs, fmt.Errorf(
+					"failed to export metrics for interval %s: %w", ivl, err),
+				)
+			}
+		}
+		if len(errs) > 0 {
+			return fmt.Errorf("failed while running final export: %w", errors.Join(errs...))
+		}
+		if err := p.db.Close(); err != nil {
+			return fmt.Errorf("failed to close database: %w", err)
+		}
+		// All future operations are invalid after db is closed
+		p.db = nil
+	}
+	return nil
+}
+
+func (p *Processor) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: true}
+}
+
+func (p *Processor) ConsumeMetrics(ctx context.Context, md pmetric.Metrics) error {
+	var errs []error
+	v := merger.Value{Metrics: pmetric.NewMetrics()}
+	md.ResourceMetrics().RemoveIf(func(rm pmetric.ResourceMetrics) bool {
+		rm.ScopeMetrics().RemoveIf(func(sm pmetric.ScopeMetrics) bool {
+			sm.Metrics().RemoveIf(func(m pmetric.Metric) bool {
+				switch t := m.Type(); t {
+				case pmetric.MetricTypeEmpty, pmetric.MetricTypeGauge, pmetric.MetricTypeSummary:
+					return false
+				case pmetric.MetricTypeSum, pmetric.MetricTypeHistogram:
+					v.MergeMetric(rm, sm, m)
+					return true
+				case pmetric.MetricTypeExponentialHistogram:
+					// TODO implement for parity with intervalprocessor
+					return false
+				default:
+					// All metric types are handled, this is unexpected
+					errs = append(errs, fmt.Errorf("unexpected metric type, dropping: %d", t))
+					return true
+				}
+			})
+			return sm.Metrics().Len() == 0
+		})
+		return rm.ScopeMetrics().Len() == 0
+	})
+
+	vb, err := v.MarshalProto()
+	if err != nil {
+		return errors.Join(append(errs, fmt.Errorf("failed to marshal value to proto binary: %w", err))...)
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	keys := make([][]byte, len(p.intervals))
+	for i, ivl := range p.intervals {
+		key := merger.NewKey(ivl, p.processingTime)
+		keys[i], err = key.Marshal()
```

Comment:
Could we pre-marshal the keys when the processing time ticks over?

Reply:
Good idea. I will give it a shot.

Reply:
Hmm, I decided to add a TODO for now. I think we should pause doing this a bit and see what ends up in the key. For example: if the tenant information ends up in the key then it might not make sense to cache things.